File: crates/haira-lexer/src/token.rs
//! Token definitions for Haira.

use logos::Logos;
use smol_str::SmolStr;

/// A token with its kind and span.
#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub kind: TokenKind,
    pub span: std::ops::Range<usize>,
}

impl Token {
    pub fn new(kind: TokenKind, span: std::ops::Range<usize>) -> Self {
        Self { kind, span }
    }
}

/// Token kinds for Haira.
#[derive(Logos, Debug, Clone, PartialEq)]
#[logos(skip r"[ \t]+")] // Skip spaces and tabs
pub enum TokenKind {
    // ========================================================================
    // Keywords
    // ========================================================================
    #[token("if")]
    If,
    #[token("else")]
    Else,
    #[token("for")]
    For,
    #[token("while")]
    While,
    #[token("return")]
    Return,
    #[token("match")]
    Match,
    #[token("true")]
    True,
    #[token("false")]
    False,
    #[token("none")]
    None,
    #[token("some")]
    Some,
    #[token("and")]
    And,
    #[token("or")]
    Or,
    #[token("not")]
    Not,
    #[token("in")]
    In,
    #[token("async")]
    Async,
    #[token("spawn")]
    Spawn,
    #[token("select")]
    Select,
    #[token("try")]
    Try,
    #[token("catch")]
    Catch,
    #[token("public")]
    Public,
    #[token("err")]
    Err,
    #[token("ok")]
    Ok,
    #[token("break")]
    Break,
    #[token("continue")]
    Continue,
    #[token("from")]
    From,
    #[token("default")]
    Default,
    #[token("ai")]
    Ai,

    // ========================================================================
    // Operators
    // ========================================================================
    #[token("+")]
    Plus,
    #[token("-")]
    Minus,
    #[token("*")]
    Star,
    #[token("/")]
    Slash,
    #[token("%")]
    Percent,

    #[token("==")]
    EqEq,
    #[token("!=")]
    Ne,
    #[token("<")]
    Lt,
    #[token(">")]
    Gt,
    #[token("<=")]
    Le,
    #[token(">=")]
    Ge,

    #[token("=")]
    Eq,
    #[token("|")]
    Pipe,
    #[token("?")]
    Question,
    #[token("=>")]
    FatArrow,
    #[token("->")]
    Arrow,
    #[token("..=")]
    DotDotEq,
    #[token("..")]
    DotDot,
    #[token(".")]
    Dot,
    #[token(":")]
    Colon,
    #[token(",")]
    Comma,
    #[token("...")]
    Ellipsis,

    // ========================================================================
    // Delimiters
    // ========================================================================
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("{")]
    LBrace,
    #[token("}")]
    RBrace,
    #[token("[")]
    LBracket,
    #[token("]")]
    RBracket,

    // ========================================================================
    // Literals
    // ========================================================================
    /// Integer literal
    #[regex(r"[0-9][0-9_]*", |lex| parse_int(lex.slice()))]
    #[regex(r"0x[0-9a-fA-F][0-9a-fA-F_]*", |lex| parse_hex(lex.slice()))]
    #[regex(r"0b[01][01_]*", |lex| parse_binary(lex.slice()))]
    #[regex(r"0o[0-7][0-7_]*", |lex| parse_octal(lex.slice()))]
    Int(i64),

    /// Float literal
    #[regex(r"[0-9][0-9_]*\.[0-9][0-9_]*", |lex| parse_float(lex.slice()))]
    Float(f64),

    /// String literal (simple strings without interpolation)
    #[regex(r#""([^"\\{]|\\.)*""#, |lex| parse_string(lex.slice()))]
    String(SmolStr),

    /// Interpolated string literal (contains `{...}` expressions)
    /// We match strings that contain `{` and parse them specially
    #[regex(r#""([^"\\]|\\.)*""#, |lex| parse_interpolated_string(lex.slice()), priority = 1)]
    InterpolatedString(SmolStr),

    /// Identifier
    #[regex(r"[a-zA-Z_][a-zA-Z0-9_]*", |lex| SmolStr::from(lex.slice()))]
    Ident(SmolStr),

    // ========================================================================
    // Whitespace and Comments
    // ========================================================================
    /// Newline (significant for statement separation)
    #[regex(r"\n|\r\n")]
    Newline,

    /// Single-line comment
    #[regex(r"//[^\n]*")]
    LineComment,

    /// Multi-line comment (handled specially)
    #[token("/*", |lex| skip_block_comment(lex))]
    BlockComment,

    /// End of file
    Eof,

    /// Error token
    Error,
}

impl TokenKind {
    /// Check if this token is a keyword.
    pub fn is_keyword(&self) -> bool {
        matches!(
            self,
            TokenKind::If
                | TokenKind::Else
                | TokenKind::For
                | TokenKind::While
                | TokenKind::Return
                | TokenKind::Match
                | TokenKind::True
                | TokenKind::False
                | TokenKind::None
                | TokenKind::Some
                | TokenKind::And
                | TokenKind::Or
                | TokenKind::Not
                | TokenKind::In
                | TokenKind::Async
                | TokenKind::Spawn
                | TokenKind::Select
                | TokenKind::Try
                | TokenKind::Catch
                | TokenKind::Public
                | TokenKind::Err
                | TokenKind::Ok
                | TokenKind::Break
                | TokenKind::Continue
                | TokenKind::From
                | TokenKind::Default
                | TokenKind::Ai
        )
    }

    /// Check if this token is a literal.
    pub fn is_literal(&self) -> bool {
        matches!(
            self,
            TokenKind::Int(_)
                | TokenKind::Float(_)
                | TokenKind::String(_)
                | TokenKind::InterpolatedString(_)
                | TokenKind::True
                | TokenKind::False
                | TokenKind::None
        )
    }

    /// Check if this token is trivia (comments, etc.)
    pub fn is_trivia(&self) -> bool {
        matches!(self, TokenKind::LineComment | TokenKind::BlockComment)
    }
}

// ============================================================================
// Helper functions for parsing
// ============================================================================

fn parse_int(s: &str) -> Option<i64> {
    let s = s.replace('_', "");
    s.parse().ok()
}

fn parse_hex(s: &str) -> Option<i64> {
    let s = s.strip_prefix("0x").unwrap_or(s).replace('_', "");
    i64::from_str_radix(&s, 16).ok()
}

fn parse_binary(s: &str) -> Option<i64> {
    let s = s.strip_prefix("0b").unwrap_or(s).replace('_', "");
    i64::from_str_radix(&s, 2).ok()
}

fn parse_octal(s: &str) -> Option<i64> {
    let s = s.strip_prefix("0o").unwrap_or(s).replace('_', "");
    i64::from_str_radix(&s, 8).ok()
}

fn parse_float(s: &str) -> Option<f64> {
    let s = s.replace('_', "");
    s.parse().ok()
}

fn parse_string(s: &str) -> Option<SmolStr> {
    // Remove quotes
    let s = s.strip_prefix('"')?.strip_suffix('"')?;

    // Handle escape sequences
    let mut result = String::with_capacity(s.len());
    let mut chars = s.chars().peekable();

    while let Some(c) = chars.next() {
        if c == '\\' {
            match chars.next() {
                Some('n') => result.push('\n'),
                Some('t') => result.push('\t'),
                Some('r') => result.push('\r'),
                Some('\\') => result.push('\\'),
                Some('"') => result.push('"'),
                Some('{') => result.push('{'),
                Some('}') => result.push('}'),
                Some(c) => {
                    result.push('\\');
                    result.push(c);
                }
                None => result.push('\\'),
            }
        } else {
            result.push(c);
        }
    }

    Some(SmolStr::from(result))
}

/// Parse an interpolated string, keeping the raw content for the parser to process.
/// Returns None if the string doesn't contain interpolation (handled by simple String).
fn parse_interpolated_string(s: &str) -> Option<SmolStr> {
    // Check if string contains unescaped `{`
    let inner = s.strip_prefix('"')?.strip_suffix('"')?;

    let mut has_interpolation = false;
    let mut chars = inner.chars().peekable();
    while let Some(c) = chars.next() {
        if c == '\\' {
            // Skip escaped character
            chars.next();
        } else if c == '{' {
            has_interpolation = true;
            break;
        }
    }

    if has_interpolation {
        // Return the raw content (without quotes) for the parser to process
        Some(SmolStr::from(inner))
    } else {
        // This shouldn't happen due to regex priority, but handle it
        None
    }
}

fn skip_block_comment(lex: &mut logos::Lexer<TokenKind>) -> logos::Skip {
    let remainder = lex.remainder();
    let mut depth = 1;
    let mut chars = remainder.char_indices();

    while let Some((i, c)) = chars.next() {
        match c {
            '*' => {
                if let Some((_, '/')) = chars.clone().next() {
                    chars.next();
                    depth -= 1;
                    if depth == 0 {
                        lex.bump(i + 2);
                        return logos::Skip;
                    }
                }
            }
            '/' => {
                if let Some((_, '*')) = chars.clone().next() {
                    chars.next();
                    depth += 1;
                }
            }
            _ => {}
        }
    }

    // Unclosed comment - bump to end
    lex.bump(remainder.len());
    logos::Skip
}

#[cfg(test)]
mod tests {
    use super::*;
    use logos::Logos;

    #[test]
    fn test_keywords() {
        let mut lex = TokenKind::lexer("if else for while return");
        assert_eq!(lex.next(), Some(Ok(TokenKind::If)));
        assert_eq!(lex.next(), Some(Ok(TokenKind::Else)));
        assert_eq!(lex.next(), Some(Ok(TokenKind::For)));
        assert_eq!(lex.next(), Some(Ok(TokenKind::While)));
        assert_eq!(lex.next(), Some(Ok(TokenKind::Return)));
    }

    #[test]
    fn test_integers() {
        let mut lex = TokenKind::lexer("42 1_000_000 0xFF 0b1010 0o755");
        assert_eq!(lex.next(), Some(Ok(TokenKind::Int(42))));
        assert_eq!(lex.next(), Some(Ok(TokenKind::Int(1_000_000))));
        assert_eq!(lex.next(), Some(Ok(TokenKind::Int(255))));
        assert_eq!(lex.next(), Some(Ok(TokenKind::Int(10))));
        assert_eq!(lex.next(), Some(Ok(TokenKind::Int(493))));
    }

    #[test]
    fn test_floats() {
        let mut lex = TokenKind::lexer("3.15 1_000.5");
        assert_eq!(lex.next(), Some(Ok(TokenKind::Float(3.15))));
        assert_eq!(lex.next(), Some(Ok(TokenKind::Float(1000.5))));
    }

    #[test]
    fn test_strings() {
        let mut lex = TokenKind::lexer(r#""hello" "world\n""#);
        assert_eq!(
            lex.next(),
            Some(Ok(TokenKind::String(SmolStr::from("hello"))))
        );
        assert_eq!(
            lex.next(),
            Some(Ok(TokenKind::String(SmolStr::from("world\n"))))
        );
    }

    #[test]
    fn test_identifiers() {
        let mut lex = TokenKind::lexer("foo bar_baz _private");
        assert_eq!(lex.next(), Some(Ok(TokenKind::Ident(SmolStr::from("foo")))));
        assert_eq!(
            lex.next(),
            Some(Ok(TokenKind::Ident(SmolStr::from("bar_baz"))))
        );
        assert_eq!(
            lex.next(),
            Some(Ok(TokenKind::Ident(SmolStr::from("_private"))))
        );
    }

    #[test]
    fn test_operators() {
        let mut lex = TokenKind::lexer("+ - * / == != <= >= = | ? => -> .. ..=");
        assert_eq!(lex.next(), Some(Ok(TokenKind::Plus)));
        assert_eq!(lex.next(), Some(Ok(TokenKind::Minus)));
        assert_eq!(lex.next(), Some(Ok(TokenKind::Star)));
        assert_eq!(lex.next(), Some(Ok(TokenKind::Slash)));
        assert_eq!(lex.next(), Some(Ok(TokenKind::EqEq)));
        assert_eq!(lex.next(), Some(Ok(TokenKind::Ne)));
        assert_eq!(lex.next(), Some(Ok(TokenKind::Le)));
        assert_eq!(lex.next(), Some(Ok(TokenKind::Ge)));
        assert_eq!(lex.next(), Some(Ok(TokenKind::Eq)));
        assert_eq!(lex.next(), Some(Ok(TokenKind::Pipe)));
        assert_eq!(lex.next(), Some(Ok(TokenKind::Question)));
        assert_eq!(lex.next(), Some(Ok(TokenKind::FatArrow)));
        assert_eq!(lex.next(), Some(Ok(TokenKind::Arrow)));
        assert_eq!(lex.next(), Some(Ok(TokenKind::DotDot)));
        assert_eq!(lex.next(), Some(Ok(TokenKind::DotDotEq)));
    }
}

File: compiler/internal/token/token.go
// Package token defines token kinds and the Token type for the Haira lexer.
package token

import "fmt"

// TokenKind represents the type of a lexical token.
type TokenKind int

const (
	// Keywords
	If TokenKind = iota
	Else
	For
	While
	Return
	Match
	True
	False
	None
	Some
	And
	Or
	Not
	In
	Async
	Spawn
	Select
	Try
	Catch
	Public
	Err
	Ok
	Break
	Continue
	From
	Default
	Provider
	Tool
	Agent
	Workflow
	Fn
	Enum
	Defer
	Import
	Step

	// Operators
	Plus      // +
	Minus     // -
	Star      // *
	Slash     // /
	Percent   // %
	EqEq      // ==
	Ne        // !=
	Lt        // <
	Gt        // >
	Le        // <=
	Ge        // >=
	Eq        // =
	PlusEq    // +=
	MinusEq   // -=
	StarEq    // *=
	SlashEq   // /=
	PercentEq // %=
	Pipe      // |
	Question  // ?
	FatArrow  // =>
	Arrow     // ->
	DotDotEq  // ..=
	DotDot    // ..
	Dot       // .
	Colon     // :
	Comma     // ,
	Ellipsis  // ...
	At        // @

	// Delimiters
	LParen   // (
	RParen   // )
	LBrace   // {
	RBrace   // }
	LBracket // [
	RBracket // ]

	// Literals
	Int                // integer literal
	Float              // float literal
	String             // string literal
	InterpolatedString // interpolated string (contains {expr})
	TripleQuoteString  // """..."""

	// Identifiers
	Ident // identifier

	// Whitespace / comments
	Newline      // \n
	LineComment  // // ...
	BlockComment // /* ... */

	// Special
	EOF   // end of file
	Error // lexer error
)

// Keywords maps keyword strings to their TokenKind.
var Keywords = map[string]TokenKind{
	"if":       If,
	"else":     Else,
	"for":      For,
	"while":    While,
	"return":   Return,
	"match":    Match,
	"true":     True,
	"false":    False,
	"none":     None,
	"some":     Some,
	"and":      And,
	"or":       Or,
	"not":      Not,
	"in":       In,
	"async":    Async,
	"spawn":    Spawn,
	"select":   Select,
	"try":      Try,
	"catch":    Catch,
	"public":   Public,
	"err":      Err,
	"ok":       Ok,
	"break":    Break,
	"continue": Continue,
	"from":     From,
	"default":  Default,
	"provider": Provider,
	"tool":     Tool,
	"agent":    Agent,
	"workflow": Workflow,
	"fn":       Fn,
	"enum":     Enum,
	"defer":    Defer,
	"import":   Import,
	"step":     Step,
}

// Token represents a lexical token with its kind, literal value, and position.
type Token struct {
	Kind  TokenKind
	Value string // raw text of the token
	Start int    // byte offset start (inclusive)
	End   int    // byte offset end (exclusive)
}

func (t Token) String() string {
	if t.Value != "" {
		return fmt.Sprintf("%s(%q)", t.Kind, t.Value)
	}
	return t.Kind.String()
}

func (k TokenKind) String() string {
	switch k {
	case If:
		return "if"
	case Else:
		return "else"
	case For:
		return "for"
	case While:
		return "while"
	case Return:
		return "return"
	case Match:
		return "match"
	case True:
		return "true"
	case False:
		return "false"
	case None:
		return "none"
	case Some:
		return "some"
	case And:
		return "and"
	case Or:
		return "or"
	case Not:
		return "not"
	case In:
		return "in"
	case Async:
		return "async"
	case Spawn:
		return "spawn"
	case Select:
		return "select"
	case Try:
		return "try"
	case Catch:
		return "catch"
	case Public:
		return "public"
	case Err:
		return "err"
	case Ok:
		return "ok"
	case Break:
		return "break"
	case Continue:
		return "continue"
	case From:
		return "from"
	case Default:
		return "default"
	case Provider:
		return "provider"
	case Tool:
		return "tool"
	case Agent:
		return "agent"
	case Workflow:
		return "workflow"
	case Fn:
		return "fn"
	case Enum:
		return "enum"
	case Defer:
		return "defer"
	case Import:
		return "import"
	case Step:
		return "step"
	case Plus:
		return "+"
	case Minus:
		return "-"
	case Star:
		return "*"
	case Slash:
		return "/"
	case Percent:
		return "%"
	case EqEq:
		return "=="
	case Ne:
		return "!="
	case Lt:
		return "<"
	case Gt:
		return ">"
	case Le:
		return "<="
	case Ge:
		return ">="
	case Eq:
		return "="
	case PlusEq:
		return "+="
	case MinusEq:
		return "-="
	case StarEq:
		return "*="
	case SlashEq:
		return "/="
	case PercentEq:
		return "%="
	case Pipe:
		return "|"
	case Question:
		return "?"
	case FatArrow:
		return "=>"
	case Arrow:
		return "->"
	case DotDotEq:
		return "..="
	case DotDot:
		return ".."
	case Dot:
		return "."
	case Colon:
		return ":"
	case Comma:
		return ","
	case Ellipsis:
		return "..."
	case At:
		return "@"
	case LParen:
		return "("
	case RParen:
		return ")"
	case LBrace:
		return "{"
	case RBrace:
		return "}"
	case LBracket:
		return "["
	case RBracket:
		return "]"
	case Int:
		return "Int"
	case Float:
		return "Float"
	case String:
		return "String"
	case InterpolatedString:
		return "InterpolatedString"
	case TripleQuoteString:
		return "TripleQuoteString"
	case Ident:
		return "Ident"
	case Newline:
		return "Newline"
	case LineComment:
		return "LineComment"
	case BlockComment:
		return "BlockComment"
	case EOF:
		return "EOF"
	case Error:
		return "Error"
	default:
		return fmt.Sprintf("TokenKind(%d)", int(k))
	}
}

// IsKeyword returns true if the token kind is a keyword.
func (k TokenKind) IsKeyword() bool {
	return k >= If && k <= Step
}

// IsLiteral returns true if the token kind is a literal value.
func (k TokenKind) IsLiteral() bool {
	return k >= Int && k <= TripleQuoteString
}

// IsTrivia returns true if the token kind is a comment.
func (k TokenKind) IsTrivia() bool {
	return k == LineComment || k == BlockComment
}
